### PR TITLE
Feature/pp 1027 implement backend report functionality

### PR DESCRIPTION
**Pull request by dbernstein**
## Description
This PR adds an AsyncTask database structure with associated facilities for pushing tasks onto a logical queue, pulling them for processing, and removing them when complete (by flagging them as complete as rather than deleting them).

Additionally it provides support for requesting for an inventory report to be generated asynchronously for a given library and user.

Finally it includes a script that runs every 15 minutes for pulling inventory report tasks off the queue, generating a CSV report for "Palace Marketplace", "BiblioBoard", "Unlimited Listens", and "Palace Bookshelf" titles, and then emailing them together to the requestor's email address.

## Motivation and Context
https://ebce-lyrasis.atlassian.net/browse/PP-1027

## Checklist

- [ ] I have updated the documentation accordingly.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by dbernstein** [comment]
@jonathangreen and @tdilauro :  for the life of me I can't figure out why the migration is failing.  I tried reworking the migration to simply add a single table with a single primary key and no other fields,  but it doesn't seem to work.  The table is being detected as removed after that upgrade operation. Strangely it works when I run against a copy of minotaur.   Do either of you guys have any suggestions how to sort this out?

**Comment by jonathangreen** [comment]
@dbernstein I believe its because your new tables are not being imported in `core/model/__init__.py`, so the migration script does not know they exist.

https://github.com/ThePalaceProject/circulation/blob/24900ec757e6aa171281525f57f8ce08b877ab94/core/model/__init__.py#L513-L516

File: core/scripts.py
Comment (dbernstein) [inline core/scripts.py:2858]: This SQL statement is big.  Could it be written using the sqlalchemy api?  Surely.  Would it be more or less readable?  I would argue that it would be less readable.  Also building the query by directly talking to the database allowed me to iterate quickly in psql rather than trying to do it iteratively in python.  So IMHO making modifications to it will be relatively easy because they can be pasted into a `psql` terminal using production data.

File: core/scripts.py
Comment (jonathangreen) [inline core/scripts.py:2772]: Since datasource is just a free form text input in the OPDS importers, I don't think hard coding the data sources here like this will accomplish what we are hoping for. Maybe this needs to be configurable, or you need to select the collection when you are generating the report. Otherwise we will be relying on the CMs to be setup with these exact strings set as the datasource.

File: core/scripts.py
Comment (jonathangreen) [reply core/scripts.py:2858]: FYI: one of the things that I don't like about sprinkling text SQL queries into our code base, is that it makes it hard when refactoring our SQLAlchemy models because the IDE can no longer tell you where particular attributes are used. 

Your point is valid here though, and I'm willing to defer to your judgment that this is the best way to approach this query.

File: api/admin/controller/report.py
Comment (jonathangreen) [inline api/admin/controller/report.py:1]: Minor: I've been trying to slowly move our code towards being able to run with mypy in strict mode. I think this code will pass those tests with almost no changes, so it would be nice to add this new file to `pyproject.toml` in the `[[tool.mypy.overrides]]` section that enables strict mode.

File: core/model/asynctask.py
Comment (jonathangreen) [inline core/model/asynctask.py:1]: Minor: I've been trying to slowly move our code towards being able to run with mypy in strict mode. I think this code will pass those tests with almost no changes, so it would be nice to add this new file to `pyproject.toml` in the `[[tool.mypy.overrides]]` section that enables strict mode.

File: core/model/asynctask.py
Comment (jonathangreen) [inline core/model/asynctask.py:26]: Minor: Same comment about `auto()` as above.

File: core/model/asynctask.py
Comment (jonathangreen) [inline core/model/asynctask.py:32]: Do we need a job of some sort to clear out old entries in this table? Maybe tasks that has finished more then a month ago or something like that. It seems like it might be a bit of a problem to have this table continue to grow forever.

File: core/model/asynctask.py
Comment (jonathangreen) [inline core/model/asynctask.py:46]: It looks like we're catching `ProblemDetailException`s in the controller, so perhaps we should be throwing a `ProblemDetailException` here. Even if we don't need them to be `ProblemDetailException` it would be nice to have these live in the `BasePalaceException` hierarchy.

**Review by jonathangreen — COMMENTED** [review]
This generally looks good @dbernstein! Added some comments to review. 

As far as functionality, my biggest concern is around using a hard coded list of datasources, since we can't rely on these always being the same. Might need a bit of discussion on the ticket about how we are finding the collections to include in this report.

File: api/admin/controller/report.py
Comment (dbernstein) [reply api/admin/controller/report.py:16]: The test in test_report seems to rely on the CirculationManagerController.  I left it in there.

File: core/model/asynctask.py
Comment (dbernstein) [reply core/model/asynctask.py]: I have a solution for this.

File: api/admin/controller/report.py
Comment (dbernstein) [reply api/admin/controller/report.py:1]: Re the SQL query:  there is a test covering it, so if there is a problem with the model, the test will fail.

File: core/model/asynctask.py
Comment (dbernstein) [reply core/model/asynctask.py]: I'll add it in.

File: api/admin/controller/report.py
Comment (jonathangreen) [reply api/admin/controller/report.py:20]:
```diff
diff --git a/api/admin/controller/__init__.py b/api/admin/controller/__init__.py
index ddf6fef3a..83eaaf4cb 100644
--- a/api/admin/controller/__init__.py
+++ b/api/admin/controller/__init__.py
@@ -64,4 +64,4 @@ def setup_admin_controllers(manager: CirculationManager):
     manager.admin_announcement_service = AnnouncementSettings(manager._db)
     manager.admin_search_controller = AdminSearchController(manager)
     manager.admin_quicksight_controller = QuickSightController(manager)
-    manager.admin_report_controller = ReportController(manager)
+    manager.admin_report_controller = ReportController(manager._db)
diff --git a/api/admin/controller/report.py b/api/admin/controller/report.py
index ccb182681..97c9463fa 100644
--- a/api/admin/controller/report.py
+++ b/api/admin/controller/report.py
@@ -4,6 +4,7 @@ from http import HTTPStatus

 import flask
 from flask import Response
+from sqlalchemy.orm import Session

 from api.controller.circulation_manager import CirculationManagerController
 from core.model import Library
@@ -14,10 +15,14 @@ from core.model.deferredtask import (
     queue_task,
 )
 from core.problem_details import INTERNAL_SERVER_ERROR
+from core.util.log import LoggerMixin
 from core.util.problem_detail import ProblemDetail


-class ReportController(CirculationManagerController):
+class ReportController(LoggerMixin):
+    def __init__(self, db: Session):
+        self._db = db
+
     def generate_inventory_report(self) -> Response | ProblemDetail:
         library: Library = getattr(flask.request, "library")
         admin: Admin = getattr(flask.request, "admin")
```

I think that patch removes the requirement on `CirculationManagerController`. There is a lot of code in `CirculationManagerController` and its parent class `BaseCirculationManagerController` that would be nice to avoid pulling into this controller if we don't need.

**Comment by dbernstein** [comment]
@jonathangreen :  thanks for the reviews on this one.  I redid the integration query,  removed the pg_advisory_lock and made a note about race conditions.

File: tests/core/test_scripts.py
Comment (dbernstein) [inline tests/core/test_scripts.py:2595]: I added this second collection/datasource  to ensure we are not picking up collections that should be out of scope.

**Review by jonathangreen — APPROVED** [review]
Looks good 🚀 

Might be nice to try the change suggested here https://github.com/ThePalaceProject/circulation/pull/1734#discussion_r1534234710 but this is good to merge either way. 

https://github.com/ThePalaceProject/circulation/pull/1739 is also approved and also has a DB migration, so whichever PR goes in first, the other one will need to resolve the conflicts with the migration ordering before merging.

**Pull request merged**
